veps/sig-compute/hypervisor-abstraction.md

```diff
@@ -0,0 +1,590 @@
+
+- **Use a separate CRD**: Devise a `KubeVirtHypervisorConfiguration` CRD. If it is found in the `kubevirt` namespace, use it. If not, fall back to `KVM`. This is an approach that meets our requirements, but we prefer to add the `HypervisorConfiguration` field following the precedent of `ArchConfiguration`.
+
+- **Querying Worker Nodes**: Iterate over worker nodes and look for a telltale device. If no special devices are found, fall back to KVM. In this approach, since the target hypervisor is not declared outright for the cluster, it would require a `Hypervisor` field in the `VMI` CRD. In this VEP, we do not have extension of `VMI` CRD in scope.
+
+- **Build-time switch**: Use build-tags to compile KubeVirt with support for a particular hypervisor. We did not choose this approach because it would require KubeVirt to build multiple versions of its components to support different hypervisors. A runtime configuration is preferable, especially given the logic for multi-hypervisor support is already in-tree.
+
+### Future Enhancements
+
+Full abstraction with—multi-device descriptors, richer domain defaults, hypervisor-driven validation, and alternate libvirt transports.
+
+- Extend `virt-handler` node-labeller to surface per-hypervisor capability labels so schedulers and operators can audit readiness without relying on device resources alone.
+
+## Scalability
+
+- Hypervisor selection resolves once per reconcile, and each component performs a constant-time lookup in its registry, keeping control-loop complexity independent of how many implementations ship.
+- Converter mutators execute in a deterministic order to avoid combinatorial growth in conditionals.
+
+## Update/Rollback Compatibility
+
+- The feature is additive. Clusters without hypervisor configuration continue to use KVM exclusively.
+- Rolling back to a version without the abstraction reverts to the existing KVM default once the new configuration field is removed or ignored.
+
+## Functional Testing Approach
+
+The goal of testing for multi-hypervisor support is to ensure that for each in-tree hypervisor implementation, all supported KubeVirt features are verified.
+
+- Unit tests for each hypervisor implementation for validating functionality concerning each extension point with comprehensive coverage for all new code.
+- Integration tests should test the following for each in-tree hypervisor:
+  - Validate if `virt-handler`'s device plugin can detect correct hypervisor device on the node and label node with correct allocatable and capacity.
+  - Validate `virt-controller`'s rendering of `virt-launcher` pod spec.
+  - Validate node-labeller's ability to add expected node labels for each hypervisor.
+  - Validate runtime adjustments to VMI, such as memlock limit adjustment and housekeeping thread management.
```

Comment:
We do not have any integration or component tests (except one small exception with the networking drivers).
AFAIK we have only e2e tests which cover anything beyond a small unit of logic.

Are you planning to add this new group of tests?